### PR TITLE
fixing the parsing and constraint generation for sbtype instructions

### PR DESCRIPTION
**Pull request by svv232**

### CODE REVIEW

File: o1vm/src/interpreters/riscv32im/interpreter.rs [hunk 1]
```diff
@@ -2111,14 +2108,52 @@ pub fn interpret_sbtype<Env: InterpreterEnv>(env: &mut Env, instr: SBInstruction
...
-    };
-    env.range_check16(&imm2, 12);
+    let imm0_12 = {
+        let imm11 = {
+            let pos = env.alloc_scratch();
+            unsafe { env.bitmask(&instruction, 8, 7, pos) }
+        };
 
-    // FIXME: check correctness of decomposition
+        env.range_check8(&imm11, 1);
```
Comment (martyall) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2117]: It seems strange that we check that a single "bit" is in range `[0,255]`, or am i misunderstanding the point?

File: o1vm/src/interpreters/riscv32im/interpreter.rs [hunk 2]
```diff
@@ -2066,9 +2066,11 @@ pub fn interpret_stype<Env: InterpreterEnv>(env: &mut Env, instr: SInstruction)
 /// Following the documentation found
 /// [here](https://www.cs.cornell.edu/courses/cs3410/2024fa/assignments/cpusim/riscv-instructions.pdf)
 pub fn interpret_sbtype<Env: InterpreterEnv>(env: &mut Env, instr: SBInstruction) {
+    /* fetch instruction pointer from the program state */
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2069]: Not sure we need these comments, but ok.

File: o1vm/src/interpreters/riscv32im/interpreter.rs (hunk 1 above)
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2117]: There should be a single `is_boolean` method.

File: o1vm/src/interpreters/riscv32im/interpreter.rs [hunk 3]
```diff
@@ -2111,14 +2108,52 @@ pub fn interpret_sbtype<Env: InterpreterEnv>(env: &mut Env, instr: SBInstruction
...
+            let pos = env.alloc_scratch();
+            unsafe { env.bitmask(&instruction, 31, 25, pos) }
+        };
+        env.range_check8(&imm5_10, 6);
+
+        let imm12 = {
+            let pos = env.alloc_scratch();
+            unsafe { env.bitmask(&instruction, 32, 31, pos) }
+        };
+        env.range_check8(&imm12, 1);
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2135]: Same: use `is_boolean`.

File: o1vm/src/interpreters/riscv32im/interpreter.rs [hunk 4]
```diff
@@ -2079,20 +2081,15 @@ pub fn interpret_sbtype<Env: InterpreterEnv>(env: &mut Env, instr: SBInstruction
             + (v1 * Env::constant(1 << 8))
             + v0
     };
+    /* fetch opcode from instruction bit 0 - 6 for a total len of 7 */
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2084]: Should we get rid of these comments?

File: o1vm/src/interpreters/riscv32im/interpreter.rs [hunk 5]
```diff
@@ -2111,14 +2108,52 @@ pub fn interpret_sbtype<Env: InterpreterEnv>(env: &mut Env, instr: SBInstruction
...
+                - (imm5_10.clone() * Env::constant(1 << 24))  // imm5_10 at bits 25-30
+                - (imm12.clone() * Env::constant(1 << 31)), // imm12 at bits 31
+        );
+
+        (imm12 * Env::constant(1 << 12))
+            + (imm11 * Env::constant(1 << 11))
+            + (imm5_10 * Env::constant(1 << 5))
+            + (imm1_4 * Env::constant(1 << 1))
+    };
+    // extra bit is because the 0th bit in the immediate is always 0 i.e you cannot jump to an odd address
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32im/interpreter.rs:2155]: I guess it is used somewhere after.

**Review by dannywillems — CHANGES_REQUESTED** [review]
See previous comment

File: o1vm/src/interpreters/riscv32im/interpreter.rs (hunk 5 above)
Comment (svv232) [reply o1vm/src/interpreters/riscv32im/interpreter.rs:2151]: yes it is used in future prs

**Review by dannywillems — APPROVED** [review]

**Review by dannywillems — APPROVED** [review]

**Pull request merged**